### DISCUSSION TITLE
[Hooks]: Add ability to run a particular code or function after all the spec files under `suites` block have been run

### DISCUSSION THREAD

**Discussion by mayankshukla94** [webdriverio/webdriverio#7664] [Q&A]
### Is your feature request related to a problem?

Currently, when adding a block of code under `afterSuite` hook, then that hook executes the specified function after each spec file. Basically it is considering the `describe` block as suite. 

However, my requirement is to run a piece of code after all the spec files specified under `suites` block have been run.

For example:
```
suites: {
    suiteA: [
      './test/specs/suiteA/specFileA.js'
    ],
    suiteB: [
      './test/specs/suiteB/specFileB.js'
    ]
  },
  ```
  
  Now I want to perform a delete operation for a cleanup task of automation data. So if that delete code I add it under `afterSuite` hook, so it executes the delete code after each spec file i.e. `specFileA.js` and `specFileB.js`

### Describe the solution you'd like.

A hook that should execute the code after all the spec files under `suites` have been run.

For example, `afterSuites` unlike `afterSuite`

### Describe alternatives you've considered.

For now, I'm using `afterSuite` hook and performing the required task by using `this.suites` in `wdio.conf.js` file. And also, using `suite` parameter of `afterSuite` block.

```
afterSuite: function (suite) {
    var lastSuiteASuiteFilePath = this.suites.suiteA[this.suites.suiteA.length - 1];
    var lastSuiteASuiteFilePathSplits = lastSuiteASuiteFilePath.split('/');
    var lastSuiteASuiteFileName = lastSuiteASuiteFilePathSplits[lastSuiteASuiteFilePathSplits.length - 1];

    var lastSuiteBSuiteFilePath = this.suites.suiteB[this.suites.suiteB.length - 1];
    var lastSuiteBSuiteFilePathSplits = lastSuiteBSuiteFilePath.split('/');
    var lastSuiteBSuiteFileName = lastSuiteBSuiteFilePathSplits[lastSuiteBSuiteFilePathSplits.length - 1];

    var suiteFilePath = suite.file;
    var suiteFilePathSplits = suiteFilePath.split('\\');
    if (suiteFilePathSplits === '') suiteFilePathSplits = suiteFilePath.split('/');
    var suiteFileName = suiteFilePathSplits[suiteFilePathSplits.length - 1];

    if (suiteFileName === lastSuiteASuiteFileName || suiteFileName === lastSuiteBSuiteFileName) {
      // code to execute after all the spec files under suites block have been run
    }
  },
  ```
  
  However, this can be a tedious task if there are lots of suites in `wdio.conf.js` file

### Additional context

No

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by christian-bromann** [comment]
> A hook that should execute the code after all the spec files under `suites` have been run.

How about the `onComplete` hook? It runs once after all spec files have been executed.

**Reply by mayankshukla94** [reply]
I tried with `OnComplete` hook but i get error `UnhandledPromiseRejectionWarning: ReferenceError: browser is not defined`.

**Reply by christian-bromann** [reply]
You don't have the browser object in `onComplete`. Do you need it for your cleanup task?

**Reply by mayankshukla94** [reply]
Yes I need browser object. Could you please help how can I do that

**Reply by christian-bromann** [reply]
You can't. Every session has its own browser instance. If you want to run a cleaning task just once for all why would you need one particular browser instance? And if so, which one of them?

**Reply by mayankshukla94** [reply]
Actually cleaning task requires to fetch one data from browser. So that's why I need browser instance.

**Reply by christian-bromann** [reply]
What kind of data?

**Reply by mayankshukla94** [reply]
Data is stored in the HTML, so we need to open that page to fetch the data.

**Reply by christian-bromann** [reply]
Then you will need to do this for every browser session, no? Which means you should do this in the `after` hook.

**Reply by mayankshukla94** [reply]
NO. Only once all the spec files i.e. at the end of the suite have run only then I need to do this.

**Reply by mayankshukla94** [reply]
hi @christian-bromann , Did you get a chance to look into it?

**Reply by christian-bromann** [reply]
no .. I really don't know what you want to achieve, sorry.

**Reply by mayankshukla94** [reply]
This is how I have defined my suite in `wdio.conf.js` file

```
suites: {
    sanity: [
      './test/specs/**/file1.js',
      './test/specs/**/file2.js',
      './test/specs/**/file3.js',
    ],
  },
 ```
 After the execution of `file3.js`, I want to run a code to perform one delete operation. The reason we cannot do it in `after` hook of `file3.js` because in future we will add more files under `sanity` suite so everytime if we update `after` hook in newly added files so that task will be tedious.

**Reply by christian-bromann** [reply]
How about then adding a cleaning test like:
```js
suites: {
    sanity: [
      './test/specs/**/file1.js',
      './test/specs/**/file2.js',
      './test/specs/**/file3.js',
      './test/specs/**/cleanup.js',
      './test/specs/**/file4.js',
    ],
  },
```

**Reply by mayankshukla94** [reply]
We already implemented the same solution, but still thought of raising an issue in case we'd have a better solution.

Thanks for the help 🙂

**Reply by christian-bromann** [reply]
What would an ideal/better solution for you look like?

**Reply by mayankshukla94** [reply]
If there was a `onSuiteComplete` hook, which also had the access to the browser instance, then that was what we were looking for.

**Reply by christian-bromann** [reply]
You have that: https://webdriver.io/docs/options#aftersuite

**Reply by mayankshukla94** [reply]
I tried `afterSuite` but it was getting executed after the end of each spec file present under `sanity` suite. However, my requirement was to execute it after the end of the `sanity` suite.
So basically, `afterSuite` hook considers one `describe` block as suite. This is what I understood after using `afterSuite` hook
